### PR TITLE
Metal Foam Turfs

### PR DESCRIPTION
**Pull request by Shadowmech88**
All metal foam reactions now create a metal foam floor if they are on a space turf. Metal foam floors are airtight, but they're flimsy and can be broken through with pretty much any tool. However, they can still serve as a quick patch for a breach, and you can build lattice and plating over them, allowing you to patch a hole without re-opening the breach.

Metal foam reactions made with aluminum no longer produce metal foam walls. If you want to make metal foam walls, you can still do it by using iron instead of aluminum.
Metal foam walls will still have metal foam floors underneath them if they were created on a space turf.

metal_foam

### CODE REVIEW

**Comment by FudgePucker** [comment]
Pretty good dude

**Comment by MercuryDoll** [comment]
> and you can build lattice and plating over them, allowing you to patch a hole without re-opening the breach

The "WE BUILD" engineers will love it.

**Comment by gbasood** [comment]
Honestly, this makes metal foam a great way to quickly patch mid-size breaches to prevent depressurization. I like it

**Comment by ghost** [comment]
Based/10 and solves the most crippling drawback to metal foams

**Comment by FudgePucker** [comment]
Now bombs might not be an automatic shuttle call, and now both aluminum and iron foam have a use

**Comment by ghost** [comment]
They never were an autocall. Too few give a fuck to fix things anymore and metal foams were great except when it triggered autism and some fuckwit just had to smash the shit out of it then not fix anything.

File: code/game/objects/effects/effect_system.dm [hunk 1]
```diff
@@ -978,6 +981,75 @@ steam.start() -- spawns the effect
...
+	if(!(locate(/obj/structure/lattice) in contents))
+		if(istype(C, /obj/item/stack/rods))
+			return
+	else if(istype(C, /obj/item/stack/tile))
+		return
+	user.delayNextAttack(10)
+	if (istype(C, /obj/item/weapon/grab))
+		var/obj/item/weapon/grab/G = C
+		G.affecting.loc = src.loc
+		for(var/mob/O in viewers(src))
```
Comment (9600bauds) [inline code/game/objects/effects/effect_system.dm]: visible_message reee

File: code/game/objects/effects/effect_system.dm [hunk 2]
```diff
@@ -978,6 +981,75 @@ steam.start() -- spawns the effect
...
+	icon_state = "foamedmetal"
+	icon_regular_floor = "foamedmetal"
+	icon_plating = "foamedmetal"
+	can_exist_under_lattice = 1
+
+/turf/simulated/floor/foamedmetal/attack_hand(mob/user as mob)
+	user.delayNextAttack(10)
+	if ((M_HULK in user.mutations) || (prob(50)))
+		to_chat(user, "<span class='notice'>You smash through \the [src].</span>")
+		for(var/mob/O in oviewers(user))
```
Comment (PJB3005) [inline code/game/objects/effects/effect_system.dm]: Visible message.

File: code/game/objects/effects/effect_system.dm [hunk 3]
```diff
@@ -978,6 +981,75 @@ steam.start() -- spawns the effect
...
+			if ((O.client && !( O.blinded )))
+				to_chat(O, "<span class='warning'>[user] smashes through \the [src].</span>")
+		src.ChangeTurf(get_base_turf(src.z))
+	else
+		to_chat(user, "<span class='notice'>You hit \the [src] but bounce off it.</span>")
+	return
```
Comment (PJB3005) [inline code/game/objects/effects/effect_system.dm]: Trailing return.

File: code/game/objects/effects/effect_system.dm [hunk 4]
```diff
@@ -978,6 +981,75 @@ steam.start() -- spawns the effect
...
+	user.delayNextAttack(10)
+	if (istype(C, /obj/item/weapon/grab))
+		var/obj/item/weapon/grab/G = C
+		G.affecting.loc = src.loc
+		for(var/mob/O in viewers(src))
+			if (O.client)
+				to_chat(O, "<span class='warning'>[G.assailant] smashes [G.affecting] through \the [src].</span>")
+		returnToPool(C)
+		src.ChangeTurf(get_base_turf(src.z))
+		return
```
Comment (PJB3005) [inline code/game/objects/effects/effect_system.dm:1005]: Trailing return.

File: code/game/objects/effects/effect_system.dm [hunk 5]
```diff
@@ -978,6 +981,75 @@ steam.start() -- spawns the effect
...
+
+	if(prob(C.force*20 - 25))
+		to_chat(user, "<span class='notice'>You smash through \the [src] with \the [C].</span>")
+		for(var/mob/O in oviewers(user))
```
Comment (PJB3005) [inline code/game/objects/effects/effect_system.dm]: Visible message.

**Comment by Zth--** [comment]
This is awesome, what about adding a few foam bombs to engineering?

File: code/game/objects/effects/effect_system.dm (hunk 4 above)
Comment (Shadowmech88) [inline code/game/objects/effects/effect_system.dm:1005]: Is it? If it gets here then the foam has been destroyed, it needs to not continue to the next if statement.

**Comment by MercuryDoll** [comment]
Yeah, would be nice to have at least one in the CE office or something.

**Comment by ghost** [comment]
Yeah, making the grenades and boxes of large and smalls in the code would be great. Large beaker in the CE locker and a box of small beakers on the eng foyer table.

**Comment by Kurfursten** [comment]
Didn't update the wiki

**Pull request merged**